src/main/telemetry/sim.c

```diff
@@ -0,0 +1,409 @@
+#include "platform.h"
+
+#if defined(USE_TELEMETRY) && defined(USE_TELEMETRY_SIM)
+
+#include <string.h>
+
+#include "common/axis.h"
+#include "common/streambuf.h"
+#include "common/utils.h"
+#include "common/printf.h"
+
+#include "drivers/time.h"
+
+#include "fc/config.h"
+#include "fc/fc_core.h"
+#include "fc/rc_controls.h"
+#include "fc/runtime_config.h"
+#include "fc/fc_init.h"
+
+#include "flight/imu.h"
+#include "flight/failsafe.h"
+#include "flight/pid.h"
+
+#include "io/gps.h"
+#include "io/serial.h"
+
+#include "navigation/navigation.h"
+#include "navigation/navigation_private.h"
+
+#include "sensors/acceleration.h"
+#include "sensors/battery.h"
+#include "sensors/boardalignment.h"
+#include "sensors/diagnostics.h"
+#include "sensors/gyro.h"
+#include "sensors/sensors.h"
+
+#include "common/string_light.h"
+#include "common/typeconversion.h"
+#include "build/debug.h"
+
+#include "telemetry/sim.h"
+#include "telemetry/telemetry.h"
+
+static serialPort_t *simPort;
+static serialPortConfig_t *portConfig;
+static bool simEnabled = false;
+
+static uint8_t atCommand[SIM_AT_COMMAND_MAX_SIZE];
+static int simTelemetryState = SIM_STATE_INIT;
+static timeMs_t sim_t_stateChange = 0;
+static uint8_t simResponse[SIM_RESPONSE_BUFFER_SIZE + 1];
+static int atCommandStatus = SIM_AT_OK;
+static bool simWaitAfterResponse = false;
+static uint8_t readState = SIM_READSTATE_RESPONSE;
+static uint8_t transmitFlags = 0;
+static timeMs_t t_lastMessageSent = 0;
+static uint8_t lastMessageTriggeredBy = 0;
+uint8_t simModuleState = SIM_MODULE_NOT_DETECTED;
+
+int simRssi;
+uint8_t accEvent = ACC_EVENT_NONE;
+char* accEventDescriptions[] = { "", "HIT! ", "DROP ", "HIT " };
+char* modeDescriptions[] = { "MAN", "ACR", "ANG", "HOR", "ALH", "POS", "RTH", "WP", "LAU", "FS" };
+const char gpsFixIndicators[] = { '!', '*', ' ' };
+
+
+bool isGroundStationNumberDefined() {
+    return telemetryConfig()->simGroundStationNumber[0] != '\0';
+}
+
+bool checkGroundStationNumber(uint8_t* rv)
+{
+    int i;
+    const uint8_t* gsn = telemetryConfig()->simGroundStationNumber;
+
+    for (i = 0; i < 16 && *gsn != '\0'; i++) gsn++;
+    if (i == 0)
+        return false;
+    for (i = 0; i < 16 && *rv != '\"'; i++) rv++;
+    gsn--; rv--;
+    for (i = 0; i < SIM_GROUND_STATION_NUMBER_DIGITS; i++) {
+        if (*rv != *gsn) return false;
+        gsn--; rv--;
+    }
+    return true;
```

Comment:
Checking from the end may not be bad but why 7 characters and why 7 is hardcoded. 
For example, my phone number is 9 characters long.

Reply:
Checking from the end is to avoid looking at the beginning which may be given in different formats, I'm getting both `0401234567` and `+358401234567`. Yes the length should be 9 characters in my case as well, but this check can fail if someone has a shorter number.

Reply:
```
number is       sms/call         stored               sms              call
empty      +358401234567  +358401234567  +358401234567 ok  0401234567 wrong

number is       sms/call         stored               sms              call
empty         0401234567     0401234567  +358401234567 wrong  0401234567 ok

comparing from the front
number is       sms/call         stored               sms              call
401234567  +358401234567      401234567  +358401234567 wrong  0401234567 wrong

number is       sms/call         stored               sms              call
401234567     0401234567      401234567  +358401234567 wrong  0401234567 wrong

comparing from the back by length(401234567)
number is       sms/call         stored               sms              call
401234567  +358401234567      401234567  +358401234567 ok     0401234567 ok

number is       sms/call         stored               sms              call
401234567     0401234567      401234567  +358401234567 ok     0401234567 ok
```
Maybe we just need a count of character from back to stored when simGroundStationNumber is empty (for example 7 +358401234567 => 1234567, 0401234567 => 1234567)
and always compare from back by length(simGroundStationNumber) count character

or maybe add simGroundStationNumberSms and compare when sms incoming
and simGroundStationNumber compare only with call incoming (it would be a solution to all problems)

Reply:
Yes we could require `sim_ground_station_number` to be given without the country code by the user, and check for that but not the first two digits...  And if it's not defined, only check the last n digits. Apparently the shortest existing mobile numbers are four digits (on the island of Niue).

Reply:
Changed this now, the check should be correct when the number is given without a country code, and when there's a + it will just ignore the first four digits (max. length of a country code).